### ISSUE TITLE
Add 'dynamic_pad' flag to `shuffle_batch` as implemented in `batch`

### ISSUE THREAD

**Issue by TimZaman** [tensorflow/tensorflow#5147]
The `tf.train.batch()` function had a nice `dynamic_pad` flag:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/input.py#L567
Is there anything holding us batch from implementing it in `tf.train.shuffle_batch()`:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/input.py#L778

**Comment by asimshankar** [comment]
@ebrevdo : What do you think?

**Comment by TimZaman** [comment]
Alternatively, is there a reason shuffle isnt just a flag for 'batch'?

**Comment by ebrevdo** [comment]
This is nontrivial because the c++ logic for dynamic_pad is implemented by adding a new queue type (PaddingFIFOQueue) in addition to the standard FIFOQueue.  The same amount of work should be required to extend the shuttle queue c++ class.  Contributions welcome.

**Comment by cbockman** [comment]
Is there a recommended way to handle this right now, strictly within tensorflow?  Or is the answer to wait for https://github.com/tensorflow/tensorflow/pull/5463 to get completed.

I.e., to use tf.train.shuffle_batch to get inputs--e.g., sentences--of varying lengths, and then pad them (as if dynamic_pad=True was passed)?  I'm unclear if the dynamic padding operation can be elegantly implemented from within TF.

**Comment by standy66** [comment]
@cbockman, I use the following workaround:

```
dtypes = list(map(lambda x: x.dtype, inputs))
shapes = list(map(lambda x: x.get_shape(), inputs))
queue = tf.RandomShuffleQueue(CAPACITY, MIN_AFTER_DEQUEUE, dtypes)
enqueue_op = queue.enqueue(inputs)
qr = tf.train.QueueRunner(queue, [enqueue_op] * NUM_THREADS)
tf.add_to_collection(tf.GraphKeys.QUEUE_RUNNERS, qr)
inputs = queue.dequeue()
for tensor, shape in zip(inputs, shapes):
    tensor.set_shape(shape)
```

Don't forget to start queue runners. If you use SparseTensors there is some extra work of packing and unpacking them into a list of plain tensors.

**Comment by cbockman** [comment]
Hi @standy66 , thanks, I may not be fully following this, but does this work when using TFRecords format w/ very large data sets which are loaded incrementally?

The above solution looks like we're loading everything into memory, which isn't viable with large data sets--but maybe I'm not interpreting it correctly.

**Comment by standy66** [comment]
@cbockman, no, we are not loading everything into memory. I use this solution myself for a research problem employing large datasets (100GB+ of raw data) serialized in TFRecords format.

I should clarify: _inputs_ is a list or tuple of tensors, representing a single training example, (in my case they are obtained by reading a TFRecord file format with TFRecorReader). Maybe this can help:

```
reader = TFRecordReader()
key, value = reader.read(filename_queue)

# deserialize is a custom function that deserializes string
# representation of a single tf.train.Example into tensors
# (features, label) representing single training example
features, label = deserialize(value)

inputs = [features, label]
...
```

I will update solution above.

**Comment by eiennohito** [comment]
@standy66 in my case, a workaround like that had a bottleneck between `ShuffleFIFOQueue` and `batch` operations. One thread over there could not keep up with the training itself.

However, increasing the number of threads of `batch` caused contention on queueing and decreased throughput compared to one thread for ~20% for even two threads. 

Sequence processing with tensorflow is not still as good as it can be.

**Comment by gunan** [comment]
Closing as a duplicate of #2354

**Comment by zh794390558** [comment]
Thanks @standy66

**Comment by parkjaewoo0611** [comment]
Thank you very much @standy66!
